ompi/mca/pml/ucx/pml_ucx.c

```diff
@@ -407,17 +411,27 @@ static ucp_ep_h mca_pml_ucx_add_proc_common(ompi_proc_t *proc)
     ucp_ep_params_t ep_params;
     ucp_address_t *address;
     ucs_status_t status;
-    ucp_ep_h ep;
+    ucp_ep_h ep = NULL;
     int ret;
 
+  check_again:
     /* Do not add a new endpoint if we already created one */
-    if (NULL != proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_PML]) {
+    if (PML_UCX_IS_VALID_ENDPOINT(proc)) {
         return proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_PML];
     }
+    if (!OPAL_ATOMIC_COMPARE_EXCHANGE_STRONG_PTR((opal_atomic_intptr_t *)&proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_PML],
+                                                 &ep, (void*)1)) {
+        /* give some slack to the other thread to create the endpooint */
+        while (!PML_UCX_IS_VALID_ENDPOINT(proc)) {
+            _opal_lifo_release_cpu();
+            opal_atomic_rmb();
+        }
+        goto check_again;
+    }
```

Comment:
Adding a proc-level lock would create the need for an UCX level proc struct instead of using the default ompi_proc_t. A lot of work for little benefit. And having a global lock would be way too inefficient.

Reply:
i think a global lock is good enough, given that mca_pml_ucx_recv_worker_addres/pmix and ucp_ep_create would probably also use global locks